go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -0,0 +1,489 @@
+							c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+								"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+							return err
+						}
+						c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
+					}
+					// TODO: Insert a reject ACL if the service doesn't have endpoints associated
+					if len(eps) == 0 {
+						klog.V(4).Infof("Service %s/%s without endpoints", name, namespace)
+					}
+				}
+			}
+			var externalIPs []string
+			//
+			// ExternalIP
+			//
+			for _, extIP := range service.Spec.ExternalIPs {
+				// only use the IPs of the same ClusterIP family
+				if utilnet.IsIPv6String(extIP) == utilnet.IsIPv6String(ip) {
+					externalIPs = append(externalIPs, extIP)
+				}
+			}
+			//
+			// LoadBalancer
+			//
+			for _, ingress := range service.Status.LoadBalancer.Ingress {
+				// only use the IPs of the same ClusterIP family
+				if ingress.IP != "" && utilnet.IsIPv6String(ingress.IP) == utilnet.IsIPv6String(ip) {
+					externalIPs = append(externalIPs, ingress.IP)
+				}
+			}
+
+			// reconcile external IPs
+			if len(externalIPs) > 0 {
+				gatewayRouters, _, err := gateway.GetOvnGateways()
+				if err != nil {
+					return err
+				}
+				for _, extIP := range externalIPs {
+					// only use the IPs of the same ClusterIP family
+					if utilnet.IsIPv6String(extIP) != utilnet.IsIPv6String(ip) {
+						continue
+					}
+					for _, gatewayRouter := range gatewayRouters {
+						lbID, err := gateway.GetGatewayLoadBalancer(gatewayRouter, svcPort.Protocol)
+						if err != nil {
+							klog.Warningf("Service Sync: Gateway router %s does not have load balancer (%v)",
+								gatewayRouter, err)
+							// TODO: why continue? should we error and requeue and retry?
+							continue
+						}
+						vip := util.JoinHostPortInt32(extIP, svcPort.Port)
+						// Reconcile OVN
+						klog.V(4).Infof("Updating ExternalIP service %s/%s with VIP %s %s", name, namespace, vip, svcPort.Protocol)
+						if err := loadbalancer.UpdateLoadBalancer(lbID, vip, eps); err != nil {
+							c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+								"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+							return err
+						}
+						c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// handlers
+
+// onServiceUpdate queues the Service for processing.
+func (c *Controller) onServiceAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Adding service %s", key)
+	c.queue.Add(key)
+}
+
+// onServiceUpdate updates the Service Selector in the cache and queues the Service for processing.
+func (c *Controller) onServiceUpdate(oldObj, newObj interface{}) {
```

Comment:
all the workqueue and informer stuff looks very similar to what @dave-tucker did. Could we consolidate this code to use the stuff he added? I'm sure this will impact his work to use workqueues so he may already have an generic solution for this.

Reply:
this code is the common boilerplate in all the kubernetes controllers, it is better and recommended to keep it in the same file

Reply:
> all the workqueue and informer stuff looks very similar to what @dave-tucker did

This code implements the kubernetes controller pattern https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md , 

```
for {
  desired := getDesiredState()
  current := getCurrentState()
  makeChanges(desired, current)
}
```
the workqueue in the factory seems to target a different purpose.

This controller is level driven, not edge driven:
- every service or endpointslice event generates a key `servicename/namespace`  
- the key is queued, the queue collapse events with the same key, so we have dedup here
- there are workers that process the queue events, since they key is unique, you don´t have 2 workers processing the same key (service in this case)
- errors are requeued, that means if in some operations we fail to update something in ovn, we´ll add the key again to the queue and we retry (this is rate limited)